rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,1165 @@
+
+Generally, this RFC opts to forego this design to avoid the major breaking changes it would introduce to current formats of types and accepted URLs. 'Implicitly' adding a constraint about additional properties is not an uncommon practice for JSON validators, and is a fairly easy step to implement when trying to validate entity properties against a given entity type.
+
+## Detecting cycles
+
+A potential complication of the proposed extension mechanism is that it allows for cycles to be created between types. This is not necessarily a problem for the type system in itself, but rather implementations.
+
+Cyclical dependencies can cause numerous issues when it comes to update logic, validation, consistency guarantees, traversal of the type tree, and more. As such, there was consideration about disallowing cyclic dependencies within the system as a conservative initial standpoint.
+
+Enforcing this constraint would perhaps be more complicated than allowing them in the first place, a cycle could happen over a very large number of types, and if implementations rely on the specification prohibiting them, then they may not be written to appropriately detect and handle external types which do not correctly abide by that restriction of the specification.
+
+An extension cycle happens when a part of an inheritance tree revisits a base URL it has already seen. As a contrived example, an entity type `Country` could be the supertype of `Region`, which in turn could be a supertype of the same `Country` entity type.
+
+It is important to note that aside from difficulties in implementations, the actual semantics of a type with a cyclical inheritance tree are well defined. As the types largely follow JSON schema semantics, constraints are composed and never overridden. Any finite cycle of types would result in a finite set of constraints, and upon detecting a cycle, an implementation can safely finish traversal once establishing the comprehensive set of constraints (therefore avoiding an infinite loop).
+
+## Disallowing unsatisfiable schemas
+
+Another uncomfortable side-effect of the extension mechanism is that it gives more routes through which an 'unsatisfiable type' can be created. An _unsatisfiable_ type is a type where it is impossible to create data that satisfies the constraints of the type.
+
+This can happen through a combination of constraints which overlap or contradict one another, where a simple example is saying a property must be both a number and an array.
+
+Prior to this proposal, the primitives over which the type system are built were limited enough to avoid being able to define conflicting constraints. However, with the introduction of type extension, it will be possible to combine two schemas which express a constraint over the same thing, and there is no guarantee that those constraints will be compatible.
```

Comment:
Applications can prevent this with their own handling logic - for example by validating that when updating the version of a type one is inheriting from, it is checked for conflicts - correct? This is simply not spec-guaranteed, if I understand correctly. @Alfred-Mountfield @TimDiekmann

Reply:
Nevermind, I read on.

Reply:
Yes. We don't enforce valid constraints as this would be very hard (if possible at all) to encapsulate in the type system itself. Applications are responsible to validate the entity types - or at least they should. If they don't the worst thing which could happen is that an entity of this type can not be created.